addon/components/course-materials.js

```diff
@@ -1,126 +1,156 @@
 import Component from '@ember/component';
-import { Promise as RSVPPromise, map } from 'rsvp';
 import EmberObject, { computed } from '@ember/object';
-import { isPresent, isEmpty } from '@ember/utils';
-import SortableTable from 'ilios-common/mixins/sortable-table';
+import { reads } from '@ember/object/computed';
+import { isEmpty, isPresent } from '@ember/utils';
+import { all, task, timeout } from 'ember-concurrency';
 import layout from '../templates/components/course-materials';
+import DS from 'ember-data';
 
-export default Component.extend(SortableTable, {
+const { PromiseArray } = DS;
+
+const DEBOUNCE_DELAY = 250;
+
+export default Component.extend({
   layout,
-  course: null,
-  sortBy: null,
+
   classNames: ['course-materials'],
-  typesWithUrl: null,
-  filter: '',
-  sessionLearningMaterials: computed('course.sessions.[]', function(){
-    const course = this.get('course');
-    return new RSVPPromise(resolve => {
-      course.get('sessions').then(sessions => {
-        map(sessions.toArray(), session => {
-          return session.get('learningMaterials');
-        }).then(learningMaterials => {
-          let flat = learningMaterials.reduce((flattened, obj) => {
-            return flattened.pushObjects(obj.toArray());
-          }, []);
-
-          resolve(flat);
-        });
-      });
-    });
+
+  clmQuery: '',
+  course: null,
+  courseSort: null,
+  sessionSort: null,
+  slmQuery: '',
+  onClmSort() {},
+  onSlmSort() {},
+  typesWithUrl: Object.freeze(['file', 'link']),
+
+  isClmLoading: reads('courseLearningMaterialObjects.isPending'),
+  isSlmLoading: reads('sessionLearningMaterialObjects.isPending'),
+
+  courseLearningMaterialObjects: computed(function() {
```

Comment:
Instead of returning a PromiseArray this should be written as ```
courseLearningMaterialObjects: computed(async function() {
    return this.fetchCourseLearningMaterials();
  }),
```

I'm also not sure why `fetchCourseLearningMaterials` has been extracted. Especially since that obscures that there should be dependent keys on `course` and `course.sessions.[]` and probably other things. Why not just include all of those in a single computed property?

Reply:
I thought about that. How would you get a loading state going while those async functions run?

Reply:
You can use the [ember-promise-helpers](https://github.com/fivetanley/ember-promise-helpers) helpers for that.  In the template `{{#if (is-pending courseLearningMaterialObjects)}}` will give you a loading block and `{{#each (await courseLearningMaterialObjects) as |obj|}} will give you access to the promise results.